allure2-model-api/src/main/java/io/qameta/allure/FileSystemResultsReader.java

```diff
@@ -0,0 +1,103 @@
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static io.qameta.allure.AllureConstants.ATTACHMENT_FILE_GLOB;
+import static io.qameta.allure.AllureConstants.TEST_RESULT_CONTAINER_FILE_GLOB;
+import static io.qameta.allure.AllureConstants.TEST_RESULT_FILE_GLOB;
+import static java.nio.file.Files.newDirectoryStream;
+
+/**
+ * @author charlie (Dmitry Baev).
+ */
+public class FileSystemResultsReader implements AllureResultsReader {
+
+    private final Path resultsDirectory;
+
+    private final ObjectMapper mapper;
+
+    private final List<ReadError> errors = new ArrayList<>();
+
+    public FileSystemResultsReader(Path resultsDirectory) {
+        this.resultsDirectory = resultsDirectory;
+        this.mapper = Allure2ModelJackson.createMapper();
+    }
+
+    @Override
+    public Stream<TestResult> readTestResults() {
+        return listFiles(resultsDirectory, TEST_RESULT_FILE_GLOB)
+                .map(this::readTestResult)
+                .filter(Optional::isPresent)
+                .map(Optional::get);
+    }
+
+    @Override
+    public Stream<TestResultContainer> readTestResultsContainers() {
+        return listFiles(resultsDirectory, TEST_RESULT_CONTAINER_FILE_GLOB)
+                .map(this::readTestResultContainer)
+                .filter(Optional::isPresent)
+                .map(Optional::get);
+    }
+
+    @Override
+    public Stream<String> findAllAttachments() {
+        return listFiles(resultsDirectory, ATTACHMENT_FILE_GLOB)
+                .map(Path::getFileName)
+                .map(Path::toString);
+    }
+
+    @Override
+    public InputStream readAttachment(String source) throws IOException {
+        return Files.newInputStream(resultsDirectory.resolve(source));
+    }
+
+    @Override
+    public List<ReadError> getErrors() {
+        return errors;
+    }
+
+    private Optional<TestResult> readTestResult(Path file) {
+        try (InputStream is = Files.newInputStream(file)) {
+            return Optional.of(mapper.readValue(is, TestResult.class));
+        } catch (IOException e) {
+            errors.add(new ReadError(e, "Could not read result file {}", file));
+            return Optional.empty();
+        }
+    }
+
+    private Optional<TestResultContainer> readTestResultContainer(Path file) {
+        try (InputStream is = Files.newInputStream(file)) {
+            return Optional.of(mapper.readValue(is, TestResultContainer.class));
+        } catch (IOException e) {
+            errors.add(new ReadError(e, "Could not read result container file {}", file));
+            return Optional.empty();
+        }
+    }
+
+    private Stream<Path> listFiles(Path directory, String glob) {
+        try (DirectoryStream<Path> directoryStream = newDirectoryStream(directory, glob)) {
+            return StreamSupport.stream(directoryStream.spliterator(), false)
+                    .filter(Files::isRegularFile)
+                    .collect(Collectors.toList())
```

Comment:
`DirectoryStream` is the lazy stream. And its gonna be closed after return. So I need to read all the files to memory, and then return it.